src/ppx/metrics.cpp

```diff
@@ -215,104 +175,112 @@ uint64_t MetricCounter::Get() const
         object["counters"] += pMetric->Export();
     }
     return object;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
-Manager::Manager()
+Run* Manager::AddRun(const std::string& name)
 {
-}
+    PPX_ASSERT_MSG(!name.empty(), "A run name must not be empty");
+    PPX_ASSERT_MSG(mRuns.find(name) == mRuns.end(), "Runs must have unique names (duplicate name detected)");
 
-Manager::~Manager()
-{
-    for (auto [name, pRun] : mRuns) {
-        delete pRun;
-    }
-    mRuns.clear();
+    auto* pRun = new Run(name);
+    auto  run  = std::unique_ptr<Run>(pRun);
+    mRuns.emplace(name, std::move(run));
+    return pRun;
 }
 
-Run* Manager::AddRun(const char* pName)
+void Manager::ExportToDisk(const std::string& baseReportName) const
 {
-    PPX_ASSERT_MSG(pName != nullptr, "A run name must not be null");
-    PPX_ASSERT_MSG(mRuns.find(pName) == mRuns.end(), "Runs must have unique names (duplicate name detected)");
+    PPX_ASSERT_MSG(!baseReportName.empty(), "Base report name must not be empty!");
 
-    Run* pRun = new Run(pName);
-    if (pRun == nullptr) {
-        return nullptr;
+    nlohmann::json content;
+    for (const auto& [name, pRun] : mRuns) {
+        content["runs"] += pRun->Export();
     }
 
-    const auto ret = mRuns.insert({pName, pRun});
-    PPX_ASSERT_MSG(ret.second, "An insertion shall always take place when adding a run");
-    return pRun;
+    // Keep the report internal to this function.
+    // Since the json library relies on strings-as-pointers rather than self-allocated objects,
+    // the lifecycle of the report is tied to the lifecycle of the runs and metrics owned by
+    // the manager. But this is opaque.
+    Report(std::move(content), baseReportName).WriteToDisk();
 }
 
-Report Manager::Export(const char* pName) const
+////////////////////////////////////////////////////////////////////////////////
+
+Manager::Report::Report(const nlohmann::json& content, const std::string& baseFilename)
+    : mContent(content)
+{
+    SetReportName(baseFilename);
+}
+
+Manager::Report::Report(nlohmann::json&& content, const std::string& baseFilename)
+    : mContent(content)
 {
-    PPX_ASSERT_NULL_ARG(pName);
+    SetReportName(baseFilename);
+}
 
-    Report          report;
-    nlohmann::json& content = report.mContent;
+void Manager::Report::WriteToDisk(bool overwriteExisting) const
+{
+    PPX_ASSERT_MSG(!mFilePath.empty(), "Filepath must not be empty!");
 
-    content["name"] = pName;
-    for (auto [name, pRun] : mRuns) {
-        content["runs"] += pRun->Export();
+    if (!overwriteExisting && std::filesystem::exists(mFilePath)) {
+        PPX_LOG_ERROR("Metrics report file cannot be written to disk. Path [" << mFilePath << "] already exists.");
+        return;
     }
-    return report;
+
+    std::filesystem::create_directories(mFilePath.parent_path());
+    std::ofstream outputFile(mFilePath.c_str(), std::ofstream::out);
```

Comment:
should you check if the file is good before writing?
I'm fine returning void and assert, as it's the standard behavior for BigWheels when we fail, but we might need to check errors.

Reply:
Oh, yeah that's probably a good call. Note that this was a copy/paste and the previous code did not check.